src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,230 @@
+>
+> 1. Let `P` be `? ToPropertyKey(V)`.
+> 2. Let `O` be `? ToObject(this value)`.
+> 3. Return `? HasOwnProperty(O, P)`.
+
+[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)
+
+and
+
+> `HasOwnProperty(O, P)`
+>
+>The abstract operation `HasOwnProperty` is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments `O` and `P` where `O` is the object and `P` is the property key. This abstract operation performs the following steps:
+>
+> 1. Assert: `Type(O)` is `Object`.
+> 2. Assert: `IsPropertyKey(P)` is `true`.
+> 3. Let `desc` be `? O.[[GetOwnProperty]](P)`.
+> 4. If `desc` is `undefined`, return `false`.
+> 5. Return `true`.
+
+[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)
+
+But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?
+
+Let's find out!
+
+## Language Types and Specification Types
+
+Let's start with something that looks familiar. The spec uses values such as `undefined`, `true`, and `false`, which we already know from JavaScript. They all are **language values**, values of **language types** which the spec also defines.
+
+The spec also uses language values internally, for example, an internal data type might contain a field whose possible values are `true` and `false`. In contrast, JavaScript engines don't typically use language values internally. For example, if the JavaScript engine is written in C++, it would typically use the C++ `true` and `false` (and not its internal representations of the JavaScript `true` and `false`).
+
+[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)
+
+In addition to language types, the spec also uses **specification types**, which are types that occur only in the spec, but not in the JavaScript language. The JavaScript engine does not need to (but is free to) implement them. In this blog post, we'll get to know the specification type Record (and its subtype Completion Record).
+
+[Spec: ECMAScript specification types](https://tc39.es/ecma262/#sec-ecmascript-specification-types)
+
+## Abstract operations
+
+**Abstract operations** are functions defined in the ECMAScript spec; they are defined for the purpose of writing the spec concisely. A JavaScript engine doesn't have to implement them as separate functions inside the engine. They cannot be directly called from JavaScript.
+
+[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)
+
+## Internal slots and internal methods
+
+**Internal slots** and **internal methods** use names enclosed in \[\[ \]\].
+
+Internal slots are data members (of a JavaScript object or a specification type). They are used for storing the state of the object. Internal methods are member functions (of a JavaScript object or a specification type).
```

Comment:
Nit: Only JS objects have internal methods. Internal methods are a pre-defined vtable that all JS objects have in the spec. Slots on the other hand are used more liberally.

Reply:
Ah, thanks for pointing out this error!